### PR TITLE
fix(@aws-amplify/datastore): introduce "settlement" guarantees to stop() and clear()

### PR DESCRIPTION
**Pull request by svidgen**
#### Description of changes

Prior to this PR, `DataStore.clear()` and `DataStore.stop` did not provide guarantees that background work was complete before returning. This caused leakage between DataStore execution contexts (e.g., stopping and clearing DataStore to update selective sync) which resulted in rogue errors and potential data corruption.

This leakage also became very evident during certain unit/interaction test cases, wherein it manifest as unstable tests.

This PR introduces a `BackgroundProcessManager` which is can be used to track lingering promises, timeouts, subscriptions, and "cleaner" functions. It provides termination signals for BG procs to listen to and terminate early if possible when a `BackgroundProcessManager` requests it. Instances of `BackgroundProcessManager` have been added to all core DataStore processes and promises, and tests have been added which test all paths &mdash; with a partial exception for subscriptions, which require more test harnessing to fully test (to be added with bug fixes in the coming weeks).

#### Reviewer guidance

1. Turn off whitespace changes.
~~2. Forgive the cruft ... _Working on it!_~~

#### Description of how you validated changes

1. Added a lot of tests.
3. Yarn linked to app and ran DataStore operations, including observes, clears, etc..

#### TODO:

- [x] fix Auth TS check
- [x] clean up bad pauses in tests
- [ ] ~propagate termination to `API.graphql()` calls~ (Deferring to a subsequent sprint/PR, as will require more complex test harnessing)
- [x] link and integ test
- [x] disabled and/or remove console logs, debugs, warns, etc. used for troubleshooting (turn prod ones into logger.debugs)
- [x] rename `this.context` -> `this.pendingWork` or `this.pendingJobs` or something more descriptive
- [x] rename or alias `exit()` to `complete()` or something more jobs/process poolish?
- [x] add an `isComplete` and use in place of `onTerminate` setting local flags where possible.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/__tests__/Retry-test.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+		let count = 0;
+
+		function suchAFailure() {
+			console.log('suchAFailure');
+			count++;
+			throw new Error('I will never succeed.');
+		}
+
+		context
+			.add(async onTermiante => retry(suchAFailure, [], () => 1, onTermiante))
```
Comment (svidgen) [inline packages/core/__tests__/Retry-test.ts:137]: typo

File: packages/core/__tests__/BackgroundProcessManager-test.ts [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+				throw new Error('Enough shenanigans!');
+			});
+			expect(true).toBe(false);
+		} catch (error) {
+			// no need to handle here.
+		}
+
+		await manager.close();
+
+		expect(true).toBe(true);
```
Comment (dpilch) [inline packages/core/__tests__/BackgroundProcessManager-test.ts:69]:
```suggestion
		expect(manager.close()).resolves.not.toThrow();
```

File: packages/core/__tests__/BackgroundProcessManager-test.ts [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+
+		// and now Jobmanager's catch handler for the rejected promise will
+		// have been invoked.
+		expect(manager.length).toBe(0);
+
+		// just demonstrating good behavior: Always close your managers.
+		await manager.close();
+	});
+
+	test('blocks new jobs once closeed', async () => {
```
Comment (dpilch) [inline packages/core/__tests__/BackgroundProcessManager-test.ts:125]: nit: typo

```suggestion
	test('blocks new jobs once closed', async () => {
```

File: packages/core/__tests__/BackgroundProcessManager-test.ts [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+		const manager = new BackgroundProcessManager();
+
+		try {
+			await manager.add(async () => {
+				throw new Error('not today, friend!');
+			});
+			expect(true).toBe(false);
+		} catch (error) {
+			expect(error.message).toEqual('not today, friend!');
+		}
```
Comment (dpilch) [inline packages/core/__tests__/BackgroundProcessManager-test.ts:52]:
```suggestion
    expect(
      manager.add(async () => {
        throw new Error('not today, friend!');
      })
    ).rejects.toThrow('not today, friend!');
```

Could use this in a few other places as well.

**Review by dpilch — COMMENTED** [review]
Haven't gotten through all of it yet, but looks good so far!

File: packages/core/src/Util/BackgroundProcessManager.ts [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+ * Creates a new scope for promises, observables, and other types of work or
+ * processes that may be running in the background. This manager provides
+ * an singular entrypoint to request termination and await completion.
+ *
+ * As work completes on its own prior to close, the manager removes them
+ * from the registry to avoid holding references to completed jobs.
+ */
+export class BackgroundProcessManager {
+	/**
+	 * `true` if no more jobs are being accepted.
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/src/Util/BackgroundProcessManager.ts:13]: Thanks for reviewing the docstrings! This is an outdated docstring. Update pending.

File: packages/core/src/Util/BackgroundProcessManager.ts [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+		// the function we call when we want to try to terminate this job.
+		let terminate;
+
+		// the promise the job can opt into listening to for termination.
+		const onTerminate = new Promise(resolve => {
+			terminate = resolve;
+		});
+
+		// finally! start the job.
+		const jobResult = job(onTerminate as any);
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/src/Util/BackgroundProcessManager.ts:156]: Good catch. I believe this is just leftover from when I was still fiddling with the interfaces. Removing.

File: packages/core/src/Util/BackgroundProcessManager.ts [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+		description?: string
+	) {
+		this.jobs.push({ promise, terminate, description });
+
+		// in all of my testing, it is safe to multi-subscribe to a promise.
+		// so, rather than create another layer of promising, we're just going
+		// to hook into the promise we already have, and when it's done
+		// (successfully or not), we no longer need to wait for it upon close.
+
+		promise
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/src/Util/BackgroundProcessManager.ts:237]: Yeah, there will be a non-measurable difference in the current use-case, but it would be better to virtually eliminate the possibility for this class to become a bottleneck.

[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/src/Util/BackgroundProcessManager.ts:237]: It actually occurred to me that I can probably just use a `Set`. Aside from performance, did you have reasons in mind to use a `Map` over a `Set`?

Some [lite benchmarking](https://www.typescriptlang.org/play?module=1&allowSyntheticDefaultImports=false#code/MYewdgzgLgBAkgFQKICUCCC4HkByBlGAXhgEYB9ABisqoG4AoegeiZigE8AHAUxj2G5gAhgCcAliDxQhUbgB4EATQAKSMnBxlkeBAD4iMJavWbtCRhx58Bw8SAMBvejBcwxYMVDFCANmIBe3ABcMAAUoACuYFAhOBEAtgBG3CIAlET6oaHphPoAbiBiACapANoAugwAvoygkLDJYMAAFvGiANYGocLxwTDQ4mAA5gA0-TaiEiH8gpMgOfpOrjB1ECA+3AB0PiBDoQDk0KJew-tjPdypDM6uq7Ai3PEgeSkQBhATdpvunt5+gaFEKgMNh8FdGMsxAAzMIPJ4vEQQbaCIZQZowACExCB6EwuDw6SWy1u4DWG02KREIBEoQARGAQLBBCAIkN0XDnq8YAB3FK8B5QCIiMDcIoY2mpG4uGpSlak2BHESwYgi7kwAAiMm42QYyyh1LCdxgHJeMBAMJNr0JsuWlp1spqyyNgiKBlVGq19qdpPWWx2ewuYwABgASByhF2bIbcKAIMS9bIwAC0-WkSqjMbjCdSqSYJCoFCqECD4JlRra7W4cFdxETuRgRJcAqFYBWIi4UBAmxEQjARRA8QAqoO4Or7WX5TzPM0ALJCTjTT4SRyyn5eXwBPrhFnRWIJZJpDINm0k+qprVu7hquecOQDdyjMILGAFYq6e0nlz6kRhDawMQGBQtBuDAchylEUDAWIADU0HWsSCFymexQGBWVZFF6iHemeMK1s+jZYcSRyyJsRTcH+2rFOChGuFUuo0eeJEQDGoTFGMULUYhMpYc2wowGgIg9uwmxQlS8ShMRWx5L4ETcBA2ScdKIz0BOZ7ctOAlCYusx2CukIeOu-xbpEu4wHESQpPhn5IdAjG8MQBFYZ4jwQCEFQwEIbxZM+r4lBU1ncYh36-jGIHEEBIFgSZkFuLB8EMUauFPkejkMZJ3yyPEbzEOlzlZSJYg+LINJ5UepVYsQHH0TRdHWcsuWZUinARBAzShFVAUQohvGtg1LnVTAVTKaptnqWieAxtptjLg5q4GX8m4hNuEF7hZh71qlLhGpJekMXlrkwO6E1QHIdb5IURTvpKWGBQhwWhH+YUwBFAFRTuMUwXBx4MTZsBJWd30-a4fX5WRFHtYphG1T9INIkIRQYR1N1dQhPX8YJQjCaJA4SdIJH7ZDQ0qbUPrksAGyiB+jQtG0IjtHSY2zvOtJjIzN7giwMDU60HQMxpGPsCzU5oppmPgtztP07SjPHULMsxlcQA) shows the `Set` outperforms a `Map`. (It probably goes without saying that *both* outperform an array. 😅) And, the change from `Array` to `Set` is trivial. So, I'll make _that_ change unless you had another compelling reason to suggest a `Map`.

And *in either case*, thanks for pointing this out! I realize pointing out performance issues can feel like premature optimization nags (and in *some* cases are), but *scalability* issues like this can bite hard *unexpectedly* down the road.

File: packages/core/src/Util/BackgroundProcessManager.ts [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+	 * Creates and registers a fabricated job for processes that need to operate
+	 * with callbacks/hooks. The returned `resolve` and `reject`
+	 * functions can be used to signal the job is done successfully or not.
+	 * The returned `onTerminate` is a promise that will resolve when the
+	 * manager is requesting the termination of the job.
+	 *
+	 * @param description Optional description to help identify pending jobs.
+	 * @returns `{ resolve, reject, onTerminate }`
+	 */
+	private addHooks(description?: string) {
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/src/Util/BackgroundProcessManager.ts:184]: Ack!

File: packages/core/src/Util/BackgroundProcessManager.ts [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+	get state() {
+		return this._state;
+	}
+
+	/**
+	 * The registered `description` of all still-pending jobs.
+	 *
+	 * @returns descriptions as an array.
+	 */
+	get pending() {
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/src/Util/BackgroundProcessManager.ts:274]: Yeah, that's a good idea. I'd like to save it though, as it would require a little rejiggering the register those child managers differently.

I also ended up not using any child managers. So, it just may not be worth the effort until/unless we end up with a manager tree down the line. YAGNI and all that.

File: packages/datastore/src/datastore/datastore.ts [hunk 10]
```diff
@@ -728,96 +770,141 @@ class DataStore {
...
-						// In Node, we need to wait for queries to be synced to prevent returning empty arrays.
-						// In the Browser, we can begin returning data once subscriptions are in place.
-						const readyType = isNode
-							? ControlMessage.SYNC_ENGINE_SYNC_QUERIES_READY
-							: ControlMessage.SYNC_ENGINE_STORAGE_SUBSCRIBED;
-
-						if (type === readyType) {
-							this.initResolve();
-						}
+				// tslint:disable-next-line:max-line-length
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/datastore/src/datastore/datastore.ts:906]: ack

File: packages/core/__tests__/BackgroundProcessManager-test.ts [hunk 11]
```diff
@@ -0,0 +1,629 @@
...
+		manager.add(async () => {
+			return new Promise(resolve => {
+				setTimeout(() => {
+					proof = true;
+					resolve();
+				}, 50);
+			});
+		});
+
+		// the job should not have completed at the time we call this.
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/__tests__/BackgroundProcessManager-test.ts:27]: No harm in adding it!

File: packages/core/__tests__/BackgroundProcessManager-test.ts (hunk 2 above)
Comment (svidgen) [reply packages/core/__tests__/BackgroundProcessManager-test.ts:69]: 👍 ... slight difference in that we always want to `await` the `close()` in testing.

File: packages/core/__tests__/BackgroundProcessManager-test.ts (hunk 4 above)
Comment (svidgen) [reply packages/core/__tests__/BackgroundProcessManager-test.ts:52]: OK. I'm not sure what I did wrong previously, but this works. I'll make a quick pass for patterns like this that are easy to update.

File: packages/core/src/Util/BackgroundProcessManager.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/src/Util/BackgroundProcessManager.ts:277]: I'm resolving this, but we can reopen if needs be.

File: packages/core/__tests__/BackgroundProcessManager-test.ts [hunk 12]
```diff
@@ -0,0 +1,619 @@
+import Observable from 'zen-observable-ts';
+import {
+	BackgroundProcessManager,
+	BackgroundProcessManagerState,
+} from '../src/Util/BackgroundProcessManager';
+
+/**
+ * NOTE: Jest's promise rejection assertion uses substring matching.
```
Comment (dpilch) [inline packages/core/__tests__/BackgroundProcessManager-test.ts:8]: We could use `toThrowErrorMatchingSnapshot` or `toThrowErrorMatchingInlineSnapshot` to avoid this if needed.

File: packages/datastore/src/datastore/datastore.ts [hunk 13]
```diff
@@ -1292,53 +1382,63 @@ class DataStore {
...
+											primaryKeysAndValues
+										);
+										message = {
+											...message,
+											element: freshElement as T,
+										};
+									}
+
+									observer.next(message as SubscriptionMessage<T>);
+								}, 'is it this one?'),
```
Comment (dpilch) [inline packages/datastore/src/datastore/datastore.ts:1425]: Missing correct description.

**Review by dpilch — COMMENTED** [review]
LGTM! A lot of great tests! A few nits; only blocker is the description on line 1425 in `datastore.ts`.

File: packages/datastore/src/datastore/datastore.ts (hunk 13 above)
Comment (svidgen) [reply packages/datastore/src/datastore/datastore.ts:1425]: 🤦

File: packages/datastore/src/sync/index.ts [hunk 14]
```diff
@@ -481,16 +517,18 @@ export class SyncEngine {
 	private syncQueriesObservable(): Observable<
 		ControlMessageType<ControlMessage>
 	> {
+		// if not online ... wait. what? what is this doing?
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/datastore/src/sync/index.ts:520]: Removing this comment for now.

**Review by dpilch — DISMISSED** [review]

File: packages/datastore/src/storage/adapter/IndexedDBAdapter.ts [hunk 15]
```diff
@@ -385,6 +386,7 @@ class IndexedDBAdapter implements Adapter {
 
 			switch (relation.relationType) {
 				case 'HAS_ONE':
+					debugger;
```
Comment (svidgen) [inline packages/datastore/src/storage/adapter/IndexedDBAdapter.ts:389]: This `debugger` line shouldn't be here.

File: packages/datastore/src/datastore/datastore.ts [hunk 16]
```diff
@@ -1581,16 +1689,16 @@ class DataStore {
 					data?.model?.name === model.name
 				) {
 					generateAndEmitSnapshot();
-					Hub.remove('api', hubCallback);
+					Hub.remove('datastore', hubCallback);
```
Comment (svidgen) [inline packages/datastore/src/datastore/datastore.ts:1692]: Review this

File: packages/datastore/__tests__/DataStore.ts [hunk 17]
```diff
@@ -35,20 +47,510 @@ const nameOf = <T>(name: keyof T) => name;
...
+			// in other words, a save from one instance does not infect another.
+			await expectIsolation(async ({ DataStore, Post, cycle }) => {
+				await DataStore.save(new Post({ title: `title from ${cycle}` }));
+				const post = await DataStore.query(Post);
+
+				expect(post.length).toEqual(1);
+				expect(post[0].title).toEqual(`title from ${cycle}`);
+
+				// try to pollute the next test
+				DataStore.save(new Post({ title: `title from ${cycle}` }));
```
Comment (iartemiev) [inline packages/datastore/__tests__/DataStore.ts:334]: Q: did this kind of pollution occur predictably without the changes in this PR? Or is this hypothetical? I.e. if I plop this test into the main branch (along with the requisite helpers) will it reliably fail?

**Review by iartemiev — APPROVED** [review]
Great work, Jon! I think this is instrumental to improving DataStore reliability. Code quality, tests, and comments are all 💯. 

I left some comments and questions but all are non-blocking.

**Comment by alharris-at** [comment]
@danrivett I think this PR was auto-closed when the parent branch was deleted. FYI @svidgen we probably need to rebase this PR against main now, I don't have permissions to reopen on this repo.

**Comment by danrivett** [comment]
Thanks @alharris-at that makes sense now, thanks for letting me know.

We were just following this PR in particular, so got notified when it got closed.

File: packages/datastore/__tests__/DataStore.ts [hunk 18]
```diff
@@ -35,20 +47,510 @@ const nameOf = <T>(name: keyof T) => name;
...
+				async ({ DataStore, Post }) =>
+					await DataStore.save(new Post({ title: 'some title' }))
+			);
+		});
+
+		test('un-awaited saves', async () => {
+			await expectIsolation(async ({ DataStore, Post }) => {
+				DataStore.save(new Post({ title: 'some title' }));
+			});
+		});
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/datastore/__tests__/DataStore.ts:153]: `expectIsolation()` makes assertions by running the test body N times back-to-back. Prior to the fix, this would cause IndexedDB state/transaction errors.

File: packages/datastore/__tests__/DataStore.ts (hunk 17 above)
Comment (svidgen) [reply packages/datastore/__tests__/DataStore.ts:334]: That's a good question. I can't recall, actually. My sense is that this particular test was an extension of the general theme where some tests' failures would show up in subsequent tests, which is super confusing. My approach in testing was to try to cover all scenarios that felt like they could cause that pollution and ensure an awaited `clear()` could guard against them. And, I would have started with the most basic, easy-to-fix cases and worked towards the complexities of observe/observeQuery, which were actually the most obvious test-blocks that would fail in cascades.

File: packages/datastore/src/sync/processors/sync.ts [hunk 19]
```diff
@@ -215,6 +228,8 @@ class SyncProcessor {
 						authToken,
 						userAgentSuffix: USER_AGENT_SUFFIX_DATASTORE,
 					});
+
+					// TODO: onTerminate.then(() => API.cancel(...))
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/datastore/src/sync/processors/sync.ts:232]: It's icing, left out to keep scope smaller. 😅 The implementation change is small, but I have a sense it would be a heavy testing burden to feel confident. And with it *looking* like a non-critical part, I left it out with a note in the PR description.

File: packages/core/__tests__/Retry-test.ts [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+		expect(returnValue).toEqual('abc');
+		expect(count).toEqual(3);
+	});
+
+	test('will not retry if a non-retryable error is returned', async () => {
+		function throwsNonRetryableError() {
+			throw new NonRetryableError('bwahahahahaha');
+		}
+
+		// TODO: how the devil do you get expect().rejects.toThrow to work here?
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/__tests__/Retry-test.ts:32]: Ohhhh ... I know what happened here. Dane helped me dig on this somewhere else and I finally realized I was red-greening these `toThrow` tests wrong. `toThrow` performs a `String.includes` match, and I falsely assumed it was an equality match. So, I was adding "not" to the end of my errors as part of the red/green cycle and not seeing an error. So, I just reverted to the non-platformy check I knew how to use sooner rather than later.

TLDR. I understand the `toThrows` behavior now. It feels a *little* dangerous. But, I'll get the relevant tests in this file updated.

File: packages/core/src/Util/BackgroundProcessManager.ts [hunk 21]
```diff
@@ -0,0 +1,409 @@
...
+		// to hook into the promise we already have, and when it's done
+		// (successfully or not), we no longer need to wait for it upon close.
+
+		promise
+			.then(() => {
+				this.jobs.delete(jobEntry);
+			})
+			.catch(() => {
+				this.jobs.delete(jobEntry);
+			});
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/src/Util/BackgroundProcessManager.ts:246]: Using `.finally` actually causes errors here. I'll have to leave a slightly hand-wavy explanation in a comment here -- it's been awhile, and I should have left a comment from the start. But, `.finally` gets called with different sequential timing, which breaks assumptions I can't remember, causes failures, etc..

Totally fair game if you want a dive on that. I should have left a good comment on this from the start.

File: packages/core/src/Util/BackgroundProcessManager.ts [hunk 22]
```diff
@@ -0,0 +1,409 @@
...
+		const onTerminate = new Promise(resolve => {
+			terminate = resolve;
+		});
+
+		// finally! start the job.
+		const jobResult = job(onTerminate);
+
+		// depending on what the job gives back, register the result
+		// so we can monitor for completion.
+		if (jobResult instanceof Promise) {
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/src/Util/BackgroundProcessManager.ts:162]: Works for me.

**Review by dpilch — COMMENTED** [review]
I think all of the CHANGELOG changes can be removed. I think they were previously on the CPK PR, but were removed at the end.

File: packages/core/__tests__/Retry-test.ts [hunk 23]
```diff
@@ -0,0 +1,225 @@
...
+
+		// try to wait until after the ~400ms retry would have occurred.
+		await new Promise(resolve => setTimeout(resolve, 300));
+
+		expect(error).toBeTruthy();
+		expect(countSnapshot).toBeGreaterThan(0);
+		expect(count).toEqual(countSnapshot);
+	});
+
+	// TODO: test expontential backoff ...
```
[the remark this answers is not in the document]
Comment (svidgen) [reply packages/core/__tests__/Retry-test.ts:223]: There might be a good path forward on this one in light of some of the time slowing stuff I ended up playing with. But, it seems noncritical. I think we're testing it indirectly through other means. Just wanted to leave a note calling out that it wasn't explicitly/directly tested, I guess.

**Review by iartemiev — APPROVED** [review]
LGTM 👍🏻

**Review by dpilch — APPROVED** [review]

**Review by katiegoines — APPROVED** [review]
@svidgen - thanks for answering my questions. Approved.

**Review by dpilch — APPROVED** [review]

**Review by iartemiev — APPROVED** [review]
LGTM 👍🏻
